### ISSUE TITLE
"Find And Replace: Select All" hangs 1.3.2 and results in unexpected selections/cursors

### ISSUE THREAD

**Issue by mnquintana** [atom/find-and-replace#627]
_From @WalterGR on December 25, 2015 4:20_

Atom 1.3.2 on OS X 10.10.5.

I have a Clojure (.clj) file whose sections are marked with the text `Section:`  I want to select all section headings, copy them, and paste them at the top of the file.

I didn't get very far, though.  Selecting `^;; .*Section:.*` seemed to hang Atom and I had to forcefully terminate it.  This reproduced reliably.

While working towards a minimal test case, I searched for `section:` instead, and found that it's more likely that it's just taking a very long time, because the resulting selections are quite numerous.  They're also rather... unexpected.

![2015-12-24 at 20 11 05 - atom selecting unexpected text](https://cloud.githubusercontent.com/assets/380880/11999785/246c4432-aa7b-11e5-8ed7-cbc6532d4e58.png)

Notice that there are a ton of resulting selections - or perhaps cursors.  I'm not entirely sure what I'm looking at, but it doesn't seem to be all instances of `section:` having been selected.

Let me know if you'd like the original file.

_Copied from original issue: atom/atom#10179_

**Comment by mnquintana** [comment]
_From @jeancroy on December 25, 2015 23:3_

There's 18 result found for section, so search is probably working well.
However there was [work](https://github.com/atom/atom/pull/9426) to change the data structure holding marker for search result. You possibly found a bug in that.

**Comment by WalterGR** [comment]
@mnquintana  @jeancroy Searching isn't the problem. It's doing "Find And Replace: Select All" afterwards.

**Comment by WalterGR** [comment]
Here are more details.

*  I can reproduce this 100%.
*  It reproduces in safe mode.
*  To restate, there are 2 issues.  Let me know if I should open the second as a separate ticket.
   *  "Find And Replace: Select All" takes an extremely long time to finish.
   *  The results of "Find And Replace: Select All" aren't as expected.
*  The time it takes to finish the operation grows much faster than the size of the document.
   *  To "Select All" after searching for `section:` in my 52 kb document, an “Atom Helper” process would consume between 14 and 20 *minutes* of CPU time on an early 2015 MacBook Pro (2x 3.1 GHz, 16 GB).
   *  To do the same in a document perhaps a third of the size, while I was seeking a minimal test case, the operation would complete in IIRC less than a minute.
*  I've attached the profiler output and 2 screenshots of the results when reproducing in the full 52 kb document.  This time it took about 15 minutes of CPU time.

<img width="1749" alt="atom-find-and-replace-ticket-627--find-and-replace-select-all-hangs-atom - profiling results - self descending" src="https://cloud.githubusercontent.com/assets/380880/12005116/5ec72c80-ab45-11e5-9f9c-9bbf6c72fb97.png">
<img width="1749" alt="atom-find-and-replace-ticket-627--find-and-replace-select-all-hangs-atom - profiling results - total descending" src="https://cloud.githubusercontent.com/assets/380880/12005117/5ec7742e-ab45-11e5-8cdf-159ae8539b97.png">
[atom-find-and-replace-ticket-627--find-and-replace-select-all-hangs-atom.cpuprofile.zip](https://github.com/atom/find-and-replace/files/72335/atom-find-and-replace-ticket-627--find-and-replace-select-all-hangs-atom.cpuprofile.zip)

**Comment by jeancroy** [comment]
Atom 1.3.2 or Atom 1.2.4 ?

**Comment by WalterGR** [comment]
@jeancroy  `atom --version` shows 1.3.2, as does the Atom menu in the app.

What you're seeing is that I installed 1.2.4 via `homebrew cask` and Atom updated itself to 1.3.2.

In other words: ignore the directory name.  :)

**Comment by warb12** [comment]
Well I am having trouble with simple search. 1.3.2 no results found for simple words I search for.
![atomsearch](https://cloud.githubusercontent.com/assets/1297978/12010705/ac9b4f50-ac7f-11e5-9dcb-f423946e90bb.png)

**Comment by adambuczynski** [comment]
I am also experiencing this since the latest Atom 1.3.2, on OS X 10.10.5. Some searches, I think particularly for finding files in the entire project, cause Atom to hang and I have to force close the window.

**Comment by WalterGR** [comment]
My specific issue no longer repros in 1.3.3.

@adambuczynski: Just to confirm, the problem you report in your comment, is it about using Find And Replace: Select All after doing a regex Find?

**Comment by adambuczynski** [comment]
No, I was doing a regular search. But I haven't experienced it anymore lately either.

**Comment by WalterGR** [comment]
@adambuczynski: Okay, thanks.

**Comment by WalterGR** [comment]
@mnquintana: Since I can no longer repro my specific issue in 1.3.3, you're welcome to close this ticket.

@warb12: Your issue is different than what this ticket is about.  You may want to open a separate ticket about your specific issue.

**Comment by WalterGR** [comment]
@50Wliu: Turns out this does still repro in 1.3.3.  Could you reopen?  Apologies.

**Comment by 50Wliu** [comment]
@WalterGR No problem!

**Comment by WalterGR** [comment]
Thanks!

Yes, still repros in 1.4.0.

**Comment by WalterGR** [comment]
The repro steps seem to be slightly different in 1.4.0.

1.  Have a file with lots (say, 108) matches of `^;;.*\* Section:.*`
2.  Open the Find and Replace panel, e.g. by pressing `cmd-f`.
3.  Enter `^;;.*\* Section:.*` as the Find text.
4.  Press the "Find" button or `cmd-g` once.
5.  Invoke "Find and Replace: Select All".

Step 4 wasn't necessary in < 1.4.0.

(It turns out that the command I was looking for is "Find and Replace: Find All" not "Select All", but this perf bug still does exist.)

**Comment by WalterGR** [comment]
Still repros in Atom 1.8.0.

**Comment by alexchandel** [comment]
Just ran into this hang. I have no idea why it's marked as needs-reproduction. The hang would be fixed by #571.

**Comment by MartinBonner** [comment]
@alexchandel The hang *might* be fixed by #571.  The problem is that it depends what you mean by "a better search engine".  *I* mean "one which supports look-around".  Trouble is, look-around can have exponential worst-case behaviour - so #571 might make thinks worse.
